python/glow/wgr/logistic_ridge_regression.py

```diff
@@ -0,0 +1,358 @@
+            response : String specifying the desired output.  Can be 'linear' to specify the direct output of the linear
+                WGR model (default) or 'sigmoid' to specify predicted label probabilities.
+
+        Returns:
+            Pandas DataFrame containing  covariate values. The shape and order match label_df such that the
+            rows are indexed by sample ID and the columns by label. The column types are float64.
+        """
+        _check_model(self.model_df)
+        _check_cv(self.cv_df)
+
+        block_prediction_df = self.reduce_block_matrix(response)
+        self.y_hat_df = flatten_prediction_df(block_prediction_df, self.sample_blocks,
+                                              self._label_df)
+
+        record_hls_event('wgrLogisticRegressionTransform')
+
+        return self.y_hat_df
+
+    def transform_loco(self, response: str = 'linear', chromosomes: List[str] = []) -> pd.DataFrame:
+        """
+        Generates predictions for the target labels in the provided label DataFrame by applying the model resulting from
+        the LogisticRidgeRegression fit method to the starting reduced block matrix using
+        a leave-one-chromosome-out (LOCO) approach.
+
+        Caches the model and cross-validation DataFrames.
+
+        Args:
+            response : String specifying the desired output.  Can be 'linear' to specify the direct output of the linear
+                WGR model (default) or 'sigmoid' to specify predicted label probabilities.
+            chromosomes : List of chromosomes for which to generate a prediction (optional). If not provided, the
+                chromosomes will be inferred from the block matrix.
+
+        Returns:
+            Pandas DataFrame containing prediction y_hat values per chromosome. The rows are indexed by sample ID and
+            chromosome; the columns are indexed by label. The column types are float64. The DataFrame is sorted using
+            chromosome as the primary sort key, and sample ID as the secondary sort key.
+        """
+        loco_chromosomes = chromosomes if chromosomes else infer_chromosomes(self.reduced_block_df)
+        loco_chromosomes.sort()
+
+        # Cache model and CV DataFrames to avoid re-computing for each chromosome
+        self.cache_model_cv_df()
+
+        y_hat_df = pd.DataFrame({})
+        orig_model_df = self.model_df
+        for chromosome in loco_chromosomes:
+            print(f"Generating predictions for chromosome {chromosome}.")
+            loco_model_df = self.model_df.filter(
+                ~f.col('header').rlike(f'^chr_{chromosome}_(alpha|block)'))
+            self.model_df = loco_model_df
+            loco_y_hat_df = self.transform(response)
+            loco_y_hat_df['contigName'] = chromosome
+            y_hat_df = y_hat_df.append(loco_y_hat_df)
+            self.model_df = orig_model_df
+
+        self.y_hat_df = y_hat_df.set_index('contigName', append=True)
+        return self.y_hat_df
```

Comment:
Actually I think it would be appropriate to uncache the model dfs and cv dfs here since we've already materialized the offsets.